### PR TITLE
Add FXIOS default state method to ScreenState

### PR DESCRIPTION
**Pull request by FilippoZazzeroni**
## :scroll: Tickets
~~[Jira ticket](https://mozilla-hub.atlassian.net/browse/FXIOS-TODO)~~
~~[Github issue](https://github.com/mozilla-mobile/firefox-ios/issues/TODO)~~

## :bulb: Description
Add `defaultState` method to `StateType` protocol from `Redux` lib. I Added directly the method to `StateType` since there are some states that don't conform to `ScreenState` but still need a default action method.

## :pencil: Checklist
You have to check all boxes before merging
- [x] Filled in the above information (tickets numbers and description of your work)
- [x] Updated the PR name to follow our [PR naming guidelines](https://github.com/mozilla-mobile/firefox-ios/wiki/Pull-Request-Naming-Guide)
- [ ] Wrote unit tests and/or ensured the tests suite is passing
- [ ] When working on UI, I checked and implemented accessibility (minimum Dynamic Text and VoiceOver)
- [ ] If needed, I updated documentation / comments for complex code and public methods
- [ ] If needed, added a backport comment (example `@Mergifyio backport release/v120`)

### CODE REVIEW

**Comment by isabelrios** [comment]
The PR is failing due to the number of warnings: https://mozilla.slack.com/archives/CAFC45W5A/p1730810775562869

File: BrowserKit/Sources/Redux/State.swift [hunk 1]
```diff
@@ -7,4 +7,10 @@ import Foundation
 /// Defines the entire application state including the UI state and any model state that you used in the app.
 /// This state is stored inside of the `Store`, then you have views and other subscribers that will get notified
 /// every single time that state updates into the entire app.
-public protocol StateType {}
+public protocol StateType {
+    /// Returns the state for a default action by clearing any transient state from previous one.
+    ///
+    /// All the state properties that have a default value into the initializer should be restore to default.
+    /// A default action could be the default case for a reducer switch case.
+    static func defaultActionState(from state: Self) -> Self
```
Comment (OrlaM) [inline BrowserKit/Sources/Redux/State.swift:15]: I don't think the word action should be used here. The state and the action are two separate things, the action triggers changes in the state but the state does not belong to the action like this method implies. I think just defaultState is descriptive enough.

File: firefox-ios/Client/Frontend/Home/Homepage Rebuild/Redux/HomepageState.swift [hunk 2]
```diff
@@ -63,11 +59,24 @@ struct HomepageState: ScreenState, Equatable {
...
+                             headerState: HeaderState.reducer(state.headerState, action),
+                             pocketState: PocketState.reducer(state.pocketState, action))
+    }
+
+    static func defaultActionState(from state: HomepageState) -> HomepageState {
+        // since the default state would depend on other reducers use `defaultActionState(for: ,action:)`
+        // as method for default action state
+        return HomepageState(
+            windowUUID: state.windowUUID,
+            headerState: state.headerState,
```
Comment (thatswinnie) [inline firefox-ios/Client/Frontend/Home/Homepage Rebuild/Redux/HomepageState.swift:78]: I think we should call the child reducers here - they should return their own default state in this case anyway and with that we wouldn't need two methods.

File: firefox-ios/Client/Frontend/Browser/BrowserViewController/State/BrowserViewControllerState.swift [hunk 3]
```diff
@@ -476,10 +468,36 @@ struct BrowserViewControllerState: ScreenState, Equatable {
...
+        )
+    }
+
+    static func defaultActionState(from state: BrowserViewControllerState) -> BrowserViewControllerState {
+        // This method should be used only in `defaultActionState(from: ,action: )` since the default state for this
+        // state depends on the action as well
+        return BrowserViewControllerState(
+            searchScreenState: state.searchScreenState,
+            showDataClearanceFlow: state.showDataClearanceFlow,
+            fakespotState: state.fakespotState,
```
Comment (thatswinnie) [inline firefox-ios/Client/Frontend/Browser/BrowserViewController/State/BrowserViewControllerState.swift:494]: As mentioned in another comment: we should call the Fakespot and Microsurvey reducers here and remove the private `defaultActionState` - which probably means that we will need to pass in the action as well.

File: BrowserKit/Sources/Redux/State.swift (hunk 1 above)
Comment (FilippoZazzeroni) [reply BrowserKit/Sources/Redux/State.swift:15]: Ok perfect thanks @OrlaM i agree i'll remove it

File: firefox-ios/Client/Frontend/Browser/BrowserViewController/State/BrowserViewControllerState.swift (hunk 3 above)
Comment (FilippoZazzeroni) [reply firefox-ios/Client/Frontend/Browser/BrowserViewController/State/BrowserViewControllerState.swift:494]: Yes true, but i think `StateType` shouldn't know about `Action` in general. i'm looking into finding another solution.

File: firefox-ios/Client/Frontend/Settings/SearchSettings/SearchSettingsState.swift [hunk 4]
```diff
@@ -12,7 +12,15 @@ struct SearchSettingsState: ScreenState, Equatable {
         self.windowUUID = windowUUID
     }
 
+    private init(windowUUID: WindowUUID) {
```
[the remark this answers is not in the document]
Comment (FilippoZazzeroni) [reply firefox-ios/Client/Frontend/Settings/SearchSettings/SearchSettingsState.swift:15]: Yes true this is not needed, i'am removing it

**Review by thatswinnie — APPROVED** [review]

**Review by OrlaM — APPROVED** [review]
LGTM

Would you mind demo'ing this to the team at the eng meeting?

**Comment by FilippoZazzeroni** [comment]
@OrlaM yes of course i'll do it 🎉

**Pull request merged**